File: 프로그래머스/lv2/84512. 모음 사전/모음 사전.py
def solution(word):
    answer = 0
    spell = 'A E I O U'.split()
    word_dict = {}
    temp = ''
    
    def add_word(word_dict, temp, length):
        if length <= 0:
            return
        
        for i in range(len(spell)):
            temp += spell[i]
            word_dict[temp] = len(word_dict)+1
            add_word(word_dict, temp, length-1)
            temp = temp[:-1]
    
    add_word(word_dict, temp, 5)
    
    return word_dict[word]

File: 프로그래머스/2/84512. 모음 사전/모음 사전.java
import java.util.*;

class Solution {
    String[] alpha = new String[]{"A", "E", "I", "O", "U"};
    Map<String, Integer> dict = new HashMap<>();
    
    public int solution(String word) {
        addWord(new StringBuilder());
        
        int answer = dict.get(word);
        return answer;
    }
    
    private void addWord(StringBuilder sb) {
        if (sb.length() == 5)
            return;
        
        for (int i = 0; i < 5; i++) {
            sb.append(alpha[i]);
            dict.put(sb.toString(), dict.size() + 1);
            addWord(sb);
            sb.deleteCharAt(sb.length() - 1);
        }
    }
}